### PR TITLE
Spark3: foundation for broadcasting CosmosClient Cache from Spark Driver and re-use the caches in the Spark Executor.

### PR DESCRIPTION
**Pull request by moderakh**
This PR provides the foundation for broadcasting cosmos client cache (for now only CollectionCache) in a spark cluster.

**importance:**
in a spark cluster spark driver JVm and spark executor JVMs running on different machines each will have a cosmos-client.
When a spark driver orchestrate a spark job, new executors will be assigned and that requires instantiating CosmosClients on the executor JVMs. CosmosClient bootstrapping consumes master resource budget and hence very quickly we will get master resource RU throttled.

This PR provides the foundation of capturing a snapshot of the CosmosClient collection cache from the spark driver and broadcasts it to all executor, hence all executors can re-use the pre-populated collection cache from driver.

For now only we support collection-cache broadcast.

TODO:
- Later we will add support for broadcasting DatabaseAccount info (GlobalEndpoint Manager)
- PartitionKeyRangeCache
- AddressCache

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosBridgeInternal.java [hunk 1]
```diff
@@ -130,6 +131,12 @@ public static CosmosClientBuilder cloneCosmosClientBuilder(CosmosClientBuilder b
         return copy;
     }
 
+    @Warning(value = INTERNAL_USE_ONLY_WARNING)
+    public static CosmosClientBuilder setUsingState(CosmosClientBuilder cosmosClientBuilder, CosmosClientState state) {
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosBridgeInternal.java:135]: Will intel sense show these contracts?

Also can it be made deprecated to start with?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import com.azure.cosmos.implementation.caches.AsyncCache;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+
+public class CosmosClientState implements Serializable {
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java:17]: Its actually metadata cache only right. 
State might mean many other aspects.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    private byte[] serializeAsyncCollectionCache(AsyncCache<String, DocumentCollection> cache) throws IOException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        ObjectOutputStream objectOutputStream = new ObjectOutputStream(baos);
+        objectOutputStream.writeObject(AsyncCache.SerializableAsyncCache.from(cache, String.class, DocumentCollection.class));
+
+        objectOutputStream.close();
+        return baos.toByteArray();
+    }
+
+    public AsyncCache<String, DocumentCollection> getCollectionInfoByNameCache() throws IOException, ClassNotFoundException {
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java]: ni: Indent

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java [hunk 4]
```diff
@@ -384,7 +387,18 @@ public void init() {
...
 
-            this.collectionCache = new RxClientCollectionCache(this, this.sessionContainer, this.gatewayProxy, this, this.retryPolicy);
+            if (state != null ) {
+                try {
+                    this.collectionCache = new RxClientCollectionCache(this, this.sessionContainer, this.gatewayProxy, this, this.retryPolicy,
+                        state.getCollectionInfoByNameCache(),
+                        state.getCollectionInfoByIdCache()
+                    );
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java]: CosmosException?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+public class CosmosClientState implements Serializable {
+    public byte[] collectionInfoByNameCache;
+    public byte[] collectionInfoByIdCache;
+
+    public CosmosClientState() {
+        System.out.println("initialized Cosmos Client State");
+    }
+
+    public void serialize(CosmosAsyncClient client) throws IOException {
+        RxDocumentClientImpl documentClient = (RxDocumentClientImpl) CosmosBridgeInternal.getAsyncDocumentClient(client);
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java]: How will be "backward compatibility" mnaged?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/RxCollectionCache.java [hunk 6]
```diff
@@ -29,9 +31,18 @@
     private final AsyncCache<String, DocumentCollection> collectionInfoByNameCache;
     private final AsyncCache<String, DocumentCollection> collectionInfoByIdCache;
 
+    public static void serialize(CosmosClientState state, RxCollectionCache cache) throws IOException {
+        state.serializeCollectionInfoByIdCache(cache.collectionInfoByIdCache);
+        state.serializeCollectionInfoByNameCache(cache.collectionInfoByNameCache);
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/RxCollectionCache.java:36]: By name is a reference or derived one. 

How about limit to primitive data only?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/RxCollectionCache.java [hunk 7]
```diff
@@ -29,9 +31,18 @@
     private final AsyncCache<String, DocumentCollection> collectionInfoByNameCache;
     private final AsyncCache<String, DocumentCollection> collectionInfoByIdCache;
 
+    public static void serialize(CosmosClientState state, RxCollectionCache cache) throws IOException {
+        state.serializeCollectionInfoByIdCache(cache.collectionInfoByIdCache);
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/RxCollectionCache.java:35]: Are routing addresses also cached?

Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/RxCollectionCache.java:35]: Saw this as TODO in description. Thanks.

---
In reply to: [545885944](https://github.com/Azure/azure-sdk-for-java/pull/18184#discussion_r545885944) [](ancestors = 545885944)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java (hunk 2 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java:17]: Good point. will change the name to reflect it is metadata caches. Thanks.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosBridgeInternal.java (hunk 1 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosBridgeInternal.java:135]: This is an established pattern in the SDK. There are many other methods in this file with the same pattern. We should follow the same pattern here.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java (hunk 3 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java]: This file is revised.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/RxCollectionCache.java (hunk 7 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/RxCollectionCache.java:35]: This PR provides the foundation for the cache broadcast but for now I only wired up the CollectionCache serialization.

Please see PR description TODOs for more info.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/RxCollectionCache.java (hunk 6 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/RxCollectionCache.java:36]: this is by a reference. serializing `cache.collectionInfoByIdCache` and `cache.collectionInfoByIdCache` are necessary for taking a snapshot of the CollectionCache.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java (hunk 5 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java]: I am assuming you are referring to serializing caches of SDK from an older version of the SDK and trying to deserialize on newer version of the SDK. 

Assuming I understood the question correctly.
1) each of the component that we serialize/deserialize has a "serialVersionUID" which ensures the serialization code and deserialization code are on the same version. If they are not Java serialization will throw exception.

see this please:
_The serialization runtime associates with each serializable class a version number, called a serialVersionUID, which is used during deserialization to verify that the sender and receiver of a serialized object have loaded classes for that object that are compatible with respect to serialization. If the receiver has loaded a class for the object that has a different serialVersionUID than that of the corresponding sender's class, then deserialization will result in an InvalidClassException. A serializable class can declare its own serialVersionUID explicitly by declaring a field named "serialVersionUID" that must be static, final, and of type long:_
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/Serializable.html

2) if this happens in a spark cluster we have a bigger problem as it means different nodes of the spark cluster are on different versions of the spark-connector, in this situation many other things will go wrong too.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java (hunk 4 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java]: thanks. addressed.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java (hunk 2 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CosmosClientState.java]: addressed

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPartitionReader.scala [hunk 8]
```diff
@@ -25,10 +30,16 @@ case class CosmosPartitionReader(config: Map[String, String], readSchema: Struct
...
   // we shouldn't recreate everytime, causing resource leak, inefficient behaviour
-  val cosmosAsyncContainer = new CosmosClientBuilder()
+  val builder = new CosmosClientBuilder()
     .endpoint(endpointConfig.endpoint)
     .key(endpointConfig.key)
-    .buildAsyncClient()
+
+  val state = cosmosClientStateHandle.value;
+
+  CosmosBridgeInternal.metadataCaches(builder, cosmosClientStateHandle.value)
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPartitionReader.scala:39]: Just my 2 cents: I don't think the Spark connector should be allowed to use internal API. Instead I think we need to add this to public surface area (as beta first). It encourages customers to use internal API if we start using it across different maven packages/binaries

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java [hunk 9]
```diff
@@ -116,6 +118,15 @@ public CosmosClientBuilder() {
         this.throttlingRetryOptions = new ThrottlingRetryOptions();
     }
 
+    CosmosClientBuilder metadataCaches(CosmosClientMetadataCachesSnapshot metadataCachesSnapshot) {
+        this.state = metadataCachesSnapshot;
+        return this;
+    }
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java:124]: This should be public instead (see comment above)

**Comment by FabianMeiswinkel** [comment]
Just a general thought - I would prefer if the changes to the *.java classes could instead be go through master - this is a rather large change and my concern is that we will run into more an d more merge conflicts in the feature branch otherwise. I also think the risk is relatively low?

**Review by FabianMeiswinkel — APPROVED** [review]
Very cool idea - like mentioned I think this should be exposed in public API instead of using internals across different Maven packages - but that shouldn't block the PR and we can discuss next year when you are back

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java (hunk 9 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java:125]: please see this commit:
https://github.com/Azure/azure-sdk-for-java/pull/18184/commits/2f31869e9c5d84b230f1664070455ef4a20f81f2

The api is internal now and also removed bridge-internal. This is a pattern we can follow elsewhere too.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosBridgeInternal.java (hunk 1 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosBridgeInternal.java]: removed the CosmosBridgeInternal completely.
see this commit please:
https://github.com/Azure/azure-sdk-for-java/commit/2f31869e9c5d84b230f1664070455ef4a20f81f2

This is a pattern to get rid of BridgeInternal classes.

**Comment by moderakh** [comment]
> Very cool idea - like mentioned I think this should be exposed in public API instead of using internals across different Maven packages - but that shouldn't block the PR and we can discuss next year when you are back


I got rid of the BridgeInternal APIs. see this commit.
Instead of adding more methods in BridgeInternal, you can follow the pattern used in this commit:

**Comment by moderakh** [comment]
closes #18237

**Pull request merged**